jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/QueuedPool.java

```diff
@@ -0,0 +1,245 @@
+        }
+        finally
+        {
+            rdLock.unlock();
+        }
+    }
+
+    @Override
+    public boolean isTerminated()
+    {
+        rdLock.lock();
+        try
+        {
+            return terminated;
+        }
+        finally
+        {
+            rdLock.unlock();
+        }
+    }
+
+    @Override
+    public Collection<Entry<P>> terminate()
+    {
+        wrLock.lock();
+        try
+        {
+            terminated = true;
+            Collection<Entry<P>> copy = new ArrayList<>(queue);
+            queue.clear();
+            return copy;
+        }
+        finally
+        {
+            wrLock.unlock();
+        }
+    }
+
+    @Override
+    public int size()
+    {
+        return queue.size();
+    }
+
+    @Override
+    public int getMaxSize()
+    {
+        return maxSize;
+    }
+
+    @Override
+    public Stream<Entry<P>> stream()
+    {
+        return queue.stream();
+    }
+
+    private static class QueuedEntry<P> implements Entry<P>
+    {
+        private final QueuedPool<P> pool;
+        //   null  -> terminated
+        //   false -> idle
+        //   true  -> in use
+        private volatile Boolean inUse = Boolean.FALSE;
+        private volatile P pooled;
+
+        private QueuedEntry(QueuedPool<P> pool)
+        {
+            this.pool = pool;
+        }
+
+        @Override
+        public boolean enable(P pooled, boolean acquire)
+        {
+            Objects.requireNonNull(pooled);
+            if (!isReserved())
+            {
+                if (pool.isTerminated())
+                    return false;
+                throw new IllegalStateException("Entry already enabled " + this + " for " + pool);
+            }
+            if (inUse == null)
+                return false;
+            this.pooled = pooled;
+
+            if (acquire)
+            {
+                if (acquire())
+                    return true;
+                this.pooled = null;
+                if (pool.isTerminated())
+                    return false;
+                throw new IllegalStateException("Entry already enabled " + this + " for " + pool);
+            }
+            return pool.requeue(this);
+        }
+
+        @Override
+        public P getPooled()
+        {
+            return pooled;
+        }
+
+        boolean acquire()
+        {
+            if (pool.isTerminated())
+                return false;
+            if (inUse == null || inUse == Boolean.TRUE)
+                return false;
+            inUse = Boolean.TRUE;
+            return true;
+        }
+
+        @Override
+        public boolean release()
+        {
+            if (inUse == null || inUse == Boolean.FALSE)
+                return false;
+            inUse = Boolean.FALSE;
+            return pool.requeue(this);
+        }
+
+        @Override
+        public boolean remove()
+        {
+            if (inUse == null)
+                return false;
+            inUse = null;
+            return true;
```

Comment:
Go back to my version with the AtomicMarkableReference.  Two volatile fields shouts race to me!  Even if it is not a race, better to use a known good pattern rather than make up our own

Reply:
Changed to use `AtomicMarkableReference`.